### PR TITLE
dataflow: improve waiting functions

### LINKED ISSUE #7175: dataflow.flex-templates.streaming_beam.e2e_test: test_flex_template_streaming_beam failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 5cbed1bcfe3aa97dbed289355cc36cfd0a66ce43
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/538cd646-2984-42a3-b1d4-d2a955b7551a), [Sponge](http://sponge2/538cd646-2984-42a3-b1d4-d2a955b7551a)
status: failed
<details><summary>Test output</summary><br><pre>Traceback (most recent call last):
  File "/workspace/dataflow/flex-templates/streaming_beam/e2e_test.py", line 100, in test_flex_template_streaming_beam
    rows = list(utils.bigquery_query(query))
  File "/workspace/dataflow/conftest.py", line 105, in bigquery_query
    for row in bigquery_client.query(query):
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/cloud/bigquery/job/query.py", line 1748, in __iter__
    return iter(self.result())
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/cloud/bigquery/job/query.py", line 1450, in result
    do_get_result()
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func
    return retry_target(
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py", line 189, in retry_target
    return target()
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/cloud/bigquery/job/query.py", line 1440, in do_get_result
    super(QueryJob, self).result(retry=retry, timeout=timeout)
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/cloud/bigquery/job/base.py", line 727, in result
    return super(_AsyncJob, self).result(timeout=timeout, **kwargs)
  File "/workspace/dataflow/flex-templates/streaming_beam/.nox/py-3-8/lib/python3.8/site-packages/google/api_core/future/polling.py", line 135, in result
    raise self._exception
google.api_core.exceptions.NotFound: 404 Not found: Table python-docs-samples-tests:dataflow_flex_templates_streaming_beam_py38_70e23b.output_table was not found in location US

(job ID: 6b46fa0d-7921-42bb-8c88-7e79dbaf947b)

                                      -----Query Job SQL Follows-----                                       

    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |
   1:SELECT * FROM python-docs-samples-tests.dataflow_flex_templates_streaming_beam_py38_70e23b.output_table
    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |</pre></details>

### PR DESCRIPTION
**Pull request by davidcavazos**
## Description

Fixes #7175, #7377

Improves the waiting functions and makes the streaming sample to wait until the output table is created.

Also replacing `subprocess.run` with `subprocess.check_call` to have better error messages. Related to #7140 and #7180

## Checklist
- [ ] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] Please **merge** this PR for me once it is approved.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

### CODE REVIEW

File: dataflow/conftest.py [hunk 1]
```diff
@@ -55,6 +55,17 @@ def hyphen_name(name: str) -> str:
...
+    @staticmethod
+    def wait_until(
+        is_done: Callable[[], bool],
+        timeout_sec: int = TIMEOUT_SEC,
+        poll_interval_sec: int = POLL_INTERVAL_SEC,
+    ) -> bool:
+        for _ in range(0, timeout_sec, poll_interval_sec):
+            if is_done():
+                return True
+        return False
```
Comment (busunkim96) [inline dataflow/conftest.py:67]: Should this function sleep for `poll_interval_sec` in the loop body?  (Or is the interval baked in somewhere else?)

Comment (davidcavazos) [reply dataflow/conftest.py:67]: True! Forgot that, thanks for keeping an eye on it :)

**Review by busunkim96 — APPROVED** [review]

**Comment by busunkim96** [comment]
Hey @dandhlee, would you mind providing a second review so it's possible to use `automerge` on this PR? Thank you!

**Comment by davidcavazos** [comment]
It looks like all tests are technically passing, but we're getting timeouts.

**Comment by dandhlee** [comment]
If the timeout issue can be fixed I'll happily approve for the 3.8 test

**Comment by davidcavazos** [comment]
Hi @dandhlee, after some discussion we decided to remove part of the test that queries the database (there were a lot of unreliable waits and makes the test run faster, from ~40 minutes to ~10 minutes). As long as the pipeline gets to a running state successfully it should be good. Writing to BigQuery should be covered in the product team's tests.

**Review by dandhlee — APPROVED** [review]
Sounds good, LGTM!

**Pull request merged**